R/convert_to_date.R

```diff
@@ -0,0 +1,151 @@
+#' Convert many date and datetime formats as may be received from Microsoft
+#' Excel
+#'
+#' @details Character conversion checks if it matches something that looks like
+#'   a Microsoft Excel numeric date, converts those to numeric, and then runs
+#'   convert_to_datetime_helper() on those numbers.  Then, character to Date or
+#'   POSIXct conversion occurs via `character_fun(x, ...)` or
+#'   `character_fun(x, tz=tz, ...)`, respectively.
+#'
+#' @param x The object to convert
+#' @param tz The timezone for POSIXct output, unless an object is POSIXt
+#'   already.  Ignored for Date output.
+#' @param ... Passed to further methods.  Eventually may be passed to
+#'   `excel_numeric_to_date()`, `base::as.POXIXct()`, or `base::as.Date()`.
+#' @param character_fun A function to convert non-numeric-looking, non-NA values
+#'   in `x` to POSIXct objects.
+#' @param allow_na_conversion Can `character_fun(x)` for the
+#'   non-numeric-looking, non-NA values return NA?
+#' @return POSIXct objects for `convert_to_datetime()` or Date objects for
+#'   `convert_to_date()`.
+#' @examples
+#' convert_to_date("2009-07-06")
+#' convert_to_date(40000)
+#' convert_to_date("40000.1")
+#' @export
+#' @family Date conversion
+#' @importFrom lubridate ymd
+convert_to_date <- function(x, ..., character_fun=lubridate::ymd, allow_na_conversion=FALSE) {
+  convert_to_datetime_helper(
+    x, ...,
+    character_fun=character_fun,
+    allow_na_conversion=allow_na_conversion,
+    out_class="Date"
+  )
+}
+
+#' @describeIn convert_to_date Convert to a date-time (POSIXct)
+#' @examples
+#' convert_to_datetime(c("2009-07-06", "40000.1", "40000", NA), character_fun=lubridate::ymd_h, truncated=1, tz="UTC")
+#' @export
+#' @importFrom lubridate ymd_hms
+convert_to_datetime <- function(x, ..., tz="UTC", character_fun=lubridate::ymd_hms, allow_na_conversion=FALSE) {
```

Comment:
`allow_na_conversion` seems to me to mean, allow NAs to be converted.  What do you think about instead, `string_conversion_failure = c("error", "warning")` or similar for that argument?  Then change @param documentation above to be something like, "if a character value fails to parse into the desired class and instead returns NA, should the function return the result with a warning or throw an error?"

Reply:
That is easier to understand.  Good suggestion!